File: tkpdhome/src/main/java/com/tokopedia/home/beranda/presentation/presenter/HomePresenter.java
package com.tokopedia.home.beranda.presentation.presenter;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.util.Log;

import com.google.gson.Gson;
import com.tokopedia.abstraction.base.view.presenter.BaseDaggerPresenter;
import com.tokopedia.core.base.adapter.Visitable;
import com.tokopedia.core.constants.DrawerActivityBroadcastReceiverConstant;
import com.tokopedia.core.constants.TokoPointDrawerBroadcastReceiverConstant;
import com.tokopedia.core.drawer2.data.viewmodel.HomeHeaderWalletAction;
import com.tokopedia.core.drawer2.data.viewmodel.TokoPointDrawerData;
import com.tokopedia.core.network.retrofit.response.ErrorHandler;
import com.tokopedia.core.shopinfo.ShopInfoActivity;
import com.tokopedia.core.shopinfo.facades.GetShopInfoRetrofit;
import com.tokopedia.core.shopinfo.models.shopmodel.ShopModel;
import com.tokopedia.core.util.DeepLinkChecker;
import com.tokopedia.core.util.PagingHandler;
import com.tokopedia.core.util.SessionHandler;
import com.tokopedia.digital.tokocash.model.CashBackData;
import com.tokopedia.home.R;
import com.tokopedia.home.beranda.domain.interactor.GetHomeDataUseCase;
import com.tokopedia.home.beranda.domain.interactor.GetLocalHomeDataUseCase;
import com.tokopedia.home.beranda.listener.HomeFeedListener;
import com.tokopedia.home.beranda.presentation.view.HomeContract;
import com.tokopedia.home.beranda.presentation.view.adapter.viewmodel.BannerViewModel;
import com.tokopedia.home.beranda.presentation.view.adapter.viewmodel.HeaderViewModel;
import com.tokopedia.home.beranda.presentation.view.subscriber.GetHomeFeedsSubscriber;
import com.tokopedia.tkpd.tkpdfeed.feedplus.domain.usecase.GetHomeFeedsUseCase;
import com.tokopedia.usecase.RequestParams;

import java.util.List;

import javax.inject.Inject;

import rx.Observable;
import rx.Subscriber;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.schedulers.Schedulers;
import rx.subscriptions.CompositeSubscription;
import rx.subscriptions.Subscriptions;

/**
 * @author by errysuprayogi on 11/27/17.
 */

public class HomePresenter extends BaseDaggerPresenter<HomeContract.View> implements HomeContract.Presenter {

    private static final String TAG = HomePresenter.class.getSimpleName();
    private static final String CURSOR_NO_NEXT_PAGE_FEED = "CURSOR_NO_NEXT_PAGE_FEED";
    private final Context context;
    protected CompositeSubscription compositeSubscription;
    protected Subscription subscription;
    @Inject
    GetLocalHomeDataUseCase localHomeDataUseCase;
    @Inject
    GetHomeDataUseCase getHomeDataUseCase;
    @Inject
    GetHomeFeedsUseCase getHomeFeedsUseCase;
    private SessionHandler sessionHandler;
    private GetShopInfoRetrofit getShopInfoRetrofit;
    private String currentCursor = "";
    private PagingHandler pagingHandler;
    private HomeFeedListener feedListener;
    private HeaderViewModel headerViewModel;
    private boolean fetchFirstData;
    private long REQUEST_DELAY = 15000;// 15s
    private long lastRequestTime;

    public HomePresenter(Context context) {
        this.context = context;
        compositeSubscription = new CompositeSubscription();
        subscription = Subscriptions.empty();
        this.pagingHandler = new PagingHandler();
        resetPageFeed();
        sessionHandler = new SessionHandler(context);
    }

    @Override
    public void detachView() {
        super.detachView();
        if (!compositeSubscription.isUnsubscribed()) {
            compositeSubscription.unsubscribe();
        }
    }

    @NonNull
    private Observable<List<Visitable>> getDataFromNetwork() {
        return getHomeDataUseCase.getExecuteObservable(RequestParams.EMPTY)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread());
    }

    @Override
    public void onFirstLaunch() {
        this.fetchFirstData = true;
    }

    @Override
    public void onResume() {
        boolean needRefresh = (lastRequestTime + REQUEST_DELAY > System.currentTimeMillis());
        if (isViewAttached() && !this.fetchFirstData && needRefresh) {
            subscription = getHomeDataUseCase.getExecuteObservable(RequestParams.EMPTY)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(new Subscriber<List<Visitable>>() {
                        @Override
                        public void onCompleted() {

                        }

                        @Override
                        public void onError(Throwable e) {

                        }

                        @Override
                        public void onNext(List<Visitable> visitables) {
                            if (isViewAttached() && isDataValid(visitables)) {
                                getView().updateListOnResume(visitables);
                                lastRequestTime = System.currentTimeMillis();
                            }
                        }
                    });
            compositeSubscription.add(subscription);
        }
    }

    @Override
    public void getHomeData() {
        if (lastRequestTime + REQUEST_DELAY > System.currentTimeMillis()){
            if(isViewAttached()){
                getView().hideLoading();
            }
        } else {
            initHeaderViewModelData();
            subscription = localHomeDataUseCase.getExecuteObservable(RequestParams.EMPTY)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .doOnNext(refreshData())
                    .onErrorResumeNext(getDataFromNetwork())
                    .subscribe(new HomeDataSubscriber());
            compositeSubscription.add(subscription);
        }
    }

    private void initHeaderViewModelData() {
        if (SessionHandler.isV4Login(context)) {
            if (headerViewModel == null) {
                headerViewModel = new HeaderViewModel();
            }
            headerViewModel.setPendingTokocashChecked(false);
        }
    }

    @NonNull
    private Action1<List<Visitable>> refreshData() {
        return new Action1<List<Visitable>>() {
            @Override
            public void call(List<Visitable> visitables) {
                compositeSubscription.add(getDataFromNetwork().subscribe(new HomeDataSubscriber()));
            }
        };
    }

    @Override
    public void updateHeaderTokoCashData(HomeHeaderWalletAction homeHeaderWalletAction) {
        if (headerViewModel == null) {
            headerViewModel = new HeaderViewModel();
        }
        headerViewModel.setWalletDataSuccess();
        headerViewModel.setHomeHeaderWalletActionData(homeHeaderWalletAction);
        getView().updateHeaderItem(headerViewModel);
    }

    @Override
    public void onHeaderTokocashErrorFromBroadcast() {
        if (headerViewModel == null) {
            headerViewModel = new HeaderViewModel();
        }

        headerViewModel.setWalletDataError();
        headerViewModel.setHomeHeaderWalletActionData(null);
        getView().updateHeaderItem(headerViewModel);
    }

    @Override
    public void updateHeaderTokoCashPendingData(CashBackData cashBackData) {
        if (headerViewModel == null) {
            headerViewModel = new HeaderViewModel();
        }
        headerViewModel.setWalletDataSuccess();
        headerViewModel.setCashBackData(cashBackData);
        headerViewModel.setPendingTokocashChecked(true);
        getView().updateHeaderItem(headerViewModel);
    }

    @Override
    public void updateHeaderTokoPointData(TokoPointDrawerData tokoPointDrawerData) {
        if (headerViewModel == null) {
            headerViewModel = new HeaderViewModel();
        }
        headerViewModel.setTokoPointDataSuccess();
        headerViewModel.setTokoPointDrawerData(tokoPointDrawerData);
        getView().updateHeaderItem(headerViewModel);
    }

    @Override
    public void onHeaderTokopointErrorFromBroadcast() {
        if (headerViewModel == null) {
            headerViewModel = new HeaderViewModel();
        }
        headerViewModel.setTokoPointDataError();
        headerViewModel.setTokoPointDrawerData(null);
        getView().updateHeaderItem(headerViewModel);
    }

    @Override
    public void onRefreshTokoPoint() {
        Intent intentGetTokoPoint = new Intent(TokoPointDrawerBroadcastReceiverConstant.INTENT_ACTION_MAIN_APP);
        if (headerViewModel == null) {
            headerViewModel = new HeaderViewModel();
        }
        headerViewModel.setTokoPointDataSuccess();
        headerViewModel.setTokoPointDrawerData(null);
        getView().updateHeaderItem(headerViewModel);
        context.sendBroadcast(intentGetTokoPoint);
    }

    @Override
    public void onRefreshTokoCash() {
        if (!SessionHandler.isV4Login(context)) return;
        Intent intentGetTokocash = new Intent(DrawerActivityBroadcastReceiverConstant.INTENT_ACTION_MAIN_APP);
        intentGetTokocash.putExtra(DrawerActivityBroadcastReceiverConstant.EXTRA_ACTION_RECEIVER,
                DrawerActivityBroadcastReceiverConstant.ACTION_RECEIVER_GET_TOKOCASH_DATA);

        if (headerViewModel == null) {
            headerViewModel = new HeaderViewModel();
        }
        headerViewModel.setWalletDataSuccess();
        headerViewModel.setHomeHeaderWalletActionData(null);
        getView().updateHeaderItem(headerViewModel);
        context.sendBroadcast(intentGetTokocash);
    }

    @Override
    public void getShopInfo(final String url, String shopDomain) {
        getShopInfoRetrofit = new GetShopInfoRetrofit(context, "", shopDomain);
        getShopInfoRetrofit.setGetShopInfoListener(new GetShopInfoRetrofit.OnGetShopInfoListener() {
            @Override
            public void onSuccess(String result) {
                if (isViewAttached()) {
                    try {
                        ShopModel shopModel = new Gson().fromJson(result, ShopModel.class);
                        if (shopModel.info != null) {
                            Bundle bundle = ShopInfoActivity.createBundle(
                                    shopModel.getInfo().getShopId(), shopModel.getInfo().getShopDomain());
                            Intent intent = new Intent(context, ShopInfoActivity.class);
                            intent.putExtras(bundle);
                            context.startActivity(intent);
                        } else {
                            getView().openWebViewURL(url, context);
                        }
                    } catch (Exception e) {
                        getView().openWebViewURL(url, context);
                    }
                }
            }

            @Override
            public void onError(String message) {
                if (isViewAttached()) {
                    getView().openWebViewURL(url, context);
                }
            }

            @Override
            public void onFailure() {
                if (isViewAttached()) {
                    getView().openWebViewURL(url, context);
                }
            }
        });
        getShopInfoRetrofit.getShopInfo();
    }

    @Override
    public void openProductPageIfValid(final String url, String shopDomain) {
        getShopInfoRetrofit = new GetShopInfoRetrofit(context, "", shopDomain);
        getShopInfoRetrofit.setGetShopInfoListener(new GetShopInfoRetrofit.OnGetShopInfoListener() {
            @Override
            public void onSuccess(String result) {
                if (isViewAttached()) {
                    try {
                        ShopModel shopModel = new Gson().fromJson(result,
                                ShopModel.class);
                        if (shopModel.info != null) {
                            DeepLinkChecker.openProduct(url, context);
                        } else {
                            getView().openWebViewURL(url, context);
                        }
                    } catch (Exception e) {
                        getView().openWebViewURL(url, context);
                    }
                }
            }

            @Override
            public void onError(String message) {
                if (isViewAttached()) {
                    getView().openWebViewURL(url, context);
                }
            }

            @Override
            public void onFailure() {
                if (isViewAttached()) {
                    getView().openWebViewURL(url, context);
                }
            }
        });
        getShopInfoRetrofit.getShopInfo();
    }

    public void getHeaderData(boolean initialStart) {
        if (!SessionHandler.isV4Login(context)) return;
        Intent intentGetTokocash = new Intent(DrawerActivityBroadcastReceiverConstant.INTENT_ACTION_MAIN_APP);
        intentGetTokocash.putExtra(DrawerActivityBroadcastReceiverConstant.EXTRA_ACTION_RECEIVER,
                DrawerActivityBroadcastReceiverConstant.ACTION_RECEIVER_GET_TOKOCASH_DATA);

        Intent intentGetTokoPoint = new Intent(TokoPointDrawerBroadcastReceiverConstant.INTENT_ACTION_MAIN_APP);

        if (initialStart && headerViewModel != null) {
            if (headerViewModel.getHomeHeaderWalletActionData() == null)
                context.sendBroadcast(intentGetTokocash);
            if (headerViewModel.getTokoPointDrawerData() == null)
                context.sendBroadcast(intentGetTokoPoint);
        } else {
            context.sendBroadcast(intentGetTokocash);
            context.sendBroadcast(intentGetTokoPoint);
        }
    }

    public void setFeedListener(HomeFeedListener feedListener) {
        this.feedListener = feedListener;
    }

    public void resetPageFeed() {
        currentCursor = "";
        pagingHandler.setPage(0);
        if (getHomeFeedsUseCase != null) {
            getHomeFeedsUseCase.unsubscribe();
        }
    }

    public void fetchNextPageFeed() {
        pagingHandler.nextPage();
        fetchCurrentPageFeed();
    }

    public void fetchCurrentPageFeed() {
        if (currentCursor == null)
            return;
        getHomeFeedsUseCase.execute(
                getHomeFeedsUseCase.getFeedPlusParam(
                        pagingHandler.getPage(),
                        sessionHandler,
                        currentCursor),
                new GetHomeFeedsSubscriber(feedListener, pagingHandler.getPage()));
    }

    public void setCursor(String currentCursor) {
        this.currentCursor = currentCursor;
    }

    public void setCursorNoNextPageFeed() {
        this.currentCursor = CURSOR_NO_NEXT_PAGE_FEED;
    }

    public boolean hasNextPageFeed() {
        return !CURSOR_NO_NEXT_PAGE_FEED.equals(currentCursor);
    }

    private class HomeDataSubscriber extends Subscriber<List<Visitable>> {

        public HomeDataSubscriber() {
        }

        @Override
        public void onStart() {
            if (isViewAttached()) {
                getView().showLoading();
            }
        }

        @Override
        public void onCompleted() {
            if (isViewAttached()) {
                getView().hideLoading();
                fetchFirstData = false;
            }
        }

        @Override
        public void onError(Throwable e) {
            if (isViewAttached()) {
                getView().showNetworkError(ErrorHandler.getErrorMessage(e));
                onCompleted();
            }
        }

        @Override
        public void onNext(List<Visitable> visitables) {
            if (isViewAttached()) {
                if (SessionHandler.isV4Login(context) && headerViewModel != null) {
                    visitables.add(0, headerViewModel);
                }
                getView().setItems(visitables);
                if (visitables.size() > 0) {
                    getView().showRecomendationButton();
                }
                if (isDataValid(visitables)) {
                    getView().removeNetworkError();
                    lastRequestTime = System.currentTimeMillis();
                } else {
                    getView().showNetworkError(context.getString(R.string.msg_network_error));
                }
            }
        }

    }

    private boolean isDataValid(List<Visitable> visitables) {
        return containsInstance(visitables, BannerViewModel.class);
    }

    public <E> boolean containsInstance(List<E> list, Class<? extends E> clazz) {
        for (E e : list) {
            if (clazz.isInstance(e)) {
                return true;
            }
        }
        return false;
    }
}

File: features/discovery/home/src/main/java/com/tokopedia/home/beranda/presentation/presenter/HomePresenter.kt
package com.tokopedia.home.beranda.presentation.presenter

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.tokopedia.abstraction.base.view.presenter.BaseDaggerPresenter
import com.tokopedia.common_wallet.balance.domain.GetWalletBalanceUseCase
import com.tokopedia.common_wallet.pendingcashback.domain.GetPendingCasbackUseCase
import com.tokopedia.dynamicbanner.domain.PlayCardHomeUseCase
import com.tokopedia.graphql.data.model.GraphqlResponse
import com.tokopedia.home.beranda.data.mapper.HomeDataMapper
import com.tokopedia.home.beranda.data.model.KeywordSearchData
import com.tokopedia.home.beranda.data.model.TokopointsDrawerHomeData
import com.tokopedia.home.beranda.data.usecase.HomeUseCase
import com.tokopedia.home.beranda.data.usecase.PlayUseCase
import com.tokopedia.home.beranda.domain.interactor.*
import com.tokopedia.home.beranda.domain.model.banner.BannerSlidesModel
import com.tokopedia.home.beranda.domain.model.review.SuggestedProductReview
import com.tokopedia.home.beranda.helper.Resource
import com.tokopedia.home.beranda.helper.map
import com.tokopedia.home.beranda.presentation.view.HomeContract
import com.tokopedia.home.beranda.presentation.view.adapter.datamodel.CashBackData
import com.tokopedia.home.beranda.presentation.view.adapter.datamodel.HomeViewModel
import com.tokopedia.home.beranda.presentation.view.adapter.datamodel.static_channel.HeaderViewModel
import com.tokopedia.home.beranda.presentation.view.subscriber.*
import com.tokopedia.home.beranda.presentation.view.viewmodel.HomeHeaderWalletAction
import com.tokopedia.kotlin.extensions.coroutines.launchCatchError
import com.tokopedia.shop.common.data.source.cloud.model.ShopInfo
import com.tokopedia.shop.common.domain.interactor.GetShopInfoByDomainUseCase
import com.tokopedia.stickylogin.domain.usecase.StickyLoginUseCase
import com.tokopedia.stickylogin.internal.StickyLoginConstant
import com.tokopedia.topads.sdk.listener.ImpressionListener
import com.tokopedia.topads.sdk.utils.ImpresionTask
import com.tokopedia.usecase.RequestParams
import com.tokopedia.user.session.UserSessionInterface
import dagger.Lazy
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.isActive
import retrofit2.Response
import rx.Observable
import rx.Subscriber
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import rx.subscriptions.Subscriptions
import timber.log.Timber
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import kotlin.coroutines.CoroutineContext

class HomePresenter(private val userSession: UserSessionInterface,
                    private val getShopInfoByDomainUseCase: GetShopInfoByDomainUseCase,
                    private val coroutineDispatcher: CoroutineDispatcher,
                    private val homeUseCase: HomeUseCase,
                    private val playUseCase: PlayUseCase,
                    private val homeDataMapper: HomeDataMapper) :
        BaseDaggerPresenter<HomeContract.View?>(), HomeContract.Presenter, CoroutineScope {

    protected var compositeSubscription: CompositeSubscription
    protected var subscription: Subscription?
    private val masterJob = SupervisorJob()

    @Inject
    lateinit var getFeedTabUseCase: GetFeedTabUseCase

    @Inject
    lateinit var sendGeolocationInfoUseCase: SendGeolocationInfoUseCase

    @Inject
    lateinit var getWalletBalanceUseCase: GetWalletBalanceUseCase

    @Inject
    lateinit var getPendingCasbackUseCase: GetPendingCasbackUseCase

    @Inject
    lateinit var getHomeTokopointsDataUseCaseLazy: Lazy<GetHomeTokopointsDataUseCase>

    @Inject
    lateinit var getKeywordSearchUseCaseLazy: Lazy<GetKeywordSearchUseCase>

    @Inject
    lateinit var stickyLoginUseCase: StickyLoginUseCase

    @Inject
    lateinit var getHomeReviewSuggestedUseCase: GetHomeReviewSuggestedUseCase

    @Inject
    lateinit var dismissHomeReviewUseCase: DismissHomeReviewUseCase

    @Inject
    lateinit var playCardHomeUseCase: PlayCardHomeUseCase


    private var isCache = true

    val homeLiveData: LiveData<HomeViewModel> = homeUseCase.getHomeData().map {
        if(fetchFirstData) fetchFirstData = false
        homeDataMapper.mapToHomeViewModel(it, isCache)
    }

    private val _updateNetworkLiveData = MutableLiveData<Resource<Any>>()
    val updateNetworkLiveData: LiveData<Resource<Any>> get() = _updateNetworkLiveData

    private var currentCursor = ""
    private lateinit var headerViewModel: HeaderViewModel
    private var fetchFirstData = false
    private val REQUEST_DELAY_HOME_DATA: Long = TimeUnit.MINUTES.toMillis(3) // 3 minutes
    private val REQUEST_DELAY_SEND_GEOLOCATION = TimeUnit.HOURS.toMillis(1) // 1 hour

    override val coroutineContext: CoroutineContext
        get() = coroutineDispatcher + masterJob

    override fun detachView() {
        super.detachView()
        if (!compositeSubscription.isUnsubscribed) {
            compositeSubscription.unsubscribe()
        }
    }

    override fun onFirstLaunch() {
        fetchFirstData = true
    }

    override fun onResume() {
        val needRefresh = lastRequestTimeHomeData + REQUEST_DELAY_HOME_DATA < System.currentTimeMillis()
        val needSendGeolocationRequest = lastRequestTimeHomeData + REQUEST_DELAY_SEND_GEOLOCATION < System.currentTimeMillis()
        if (isViewAttached && !fetchFirstData && needRefresh) {
            updateHomeData()
        }
        if (needSendGeolocationRequest && view?.hasGeolocationPermission() == true) {
            view?.detectAndSendLocation()
        }
        getTokocashBalance()
        getTokopoint()
        searchHint()
        getStickyContent()
    }

    fun sendGeolocationData() {
        sendGeolocationInfoUseCase?.createObservable(RequestParams.EMPTY)
                .subscribeOn(Schedulers.io())
                .unsubscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Subscriber<Response<String>>() {
                    override fun onCompleted() {}
                    override fun onError(e: Throwable) {}
                    override fun onNext(s: Response<String>?) {
                        lastRequestTimeSendGeolocation = System.currentTimeMillis()
                    }
                })
    }

    override fun dismissReview() {
        view?.onSuccessDismissReview()

        dismissHomeReviewUseCase.execute(RequestParams.EMPTY, object: Subscriber<String>(){
            override fun onNext(t: String?) {}

            override fun onCompleted() {}

            override fun onError(e: Throwable?) {}

        })
    }

    override fun getSuggestedReview() {
        getHomeReviewSuggestedUseCase.execute(RequestParams.EMPTY, object: Subscriber<SuggestedProductReview>() {
            override fun onCompleted() {}

            override fun onError(e: Throwable?) {
                if(isViewAttached){
                    view?.onErrorGetReviewData()
                }
            }

            override fun onNext(suggestedProductReview: SuggestedProductReview?) {
                if(isViewAttached) {
                    view?.onSuccessGetReviewData(suggestedProductReview)
                }
            }
        })
    }

    override fun getHomeData() {
        initHeaderViewModelData()
        _updateNetworkLiveData.value = Resource.loading(null)
        lastRequestTimeSendGeolocation = System.currentTimeMillis()
        launchCatchError(coroutineContext, block = {
            val resource = homeUseCase.updateHomeData()
            isCache = false
            _updateNetworkLiveData.value = resource
        }){
            Timber.tag(HomePresenter::class.java.name).e(it)
            _updateNetworkLiveData.value = Resource.error(Throwable(), null)
        }
    }

    override fun updateHomeData() {
        lastRequestTimeHomeData = System.currentTimeMillis()
        launchCatchError(coroutineContext, block = {
            val resource = homeUseCase.updateHomeData()
            isCache = false
            _updateNetworkLiveData.value = resource
        }){
            Timber.tag(HomePresenter::class.java.name).e(it)
            _updateNetworkLiveData.value = Resource.error(Throwable(), null)
        }
    }

    private fun initHeaderViewModelData() {
        if (userSession.isLoggedIn) {
            getHeaderViewModel().isPendingTokocashChecked = false
        }
    }

    fun setCache(isCache: Boolean){
        this.isCache = isCache
    }

    override fun updateHeaderTokoCashData(homeHeaderWalletAction: HomeHeaderWalletAction) {
        getHeaderViewModel().setWalletDataSuccess()
        getHeaderViewModel().homeHeaderWalletActionData = homeHeaderWalletAction
        view?.updateHeaderItem(getHeaderViewModel())
    }

    override fun showPopUpIntroWalletOvo(applinkActivation: String) {
        view?.showPopupIntroOvo(applinkActivation)
    }

    override fun onHeaderTokocashError() {
        getHeaderViewModel().setWalletDataError()
        getHeaderViewModel().homeHeaderWalletActionData = null
        view?.updateHeaderItem(getHeaderViewModel())
    }

    override fun updateHeaderTokoCashPendingData(cashBackData: CashBackData) {
        getHeaderViewModel().setWalletDataSuccess()
        getHeaderViewModel().cashBackData = cashBackData
        getHeaderViewModel().isPendingTokocashChecked = true
        view?.updateHeaderItem(getHeaderViewModel())
    }

    override fun onHeaderTokopointError() {
        getHeaderViewModel().setTokoPointDataError()
        getHeaderViewModel().tokoPointDrawerData = null
        getHeaderViewModel().tokopointsDrawerHomeData = null
        view?.updateHeaderItem(getHeaderViewModel())
    }

    override fun onRefreshTokoPoint() {
        getHeaderViewModel().setTokoPointDataSuccess()
        getHeaderViewModel().tokoPointDrawerData = null
        getHeaderViewModel().tokopointsDrawerHomeData = null
        view?.updateHeaderItem(getHeaderViewModel())
        getTokopoint()
    }

    override fun onRefreshTokoCash() {
        if (!userSession.isLoggedIn) return
        getHeaderViewModel().setWalletDataSuccess()
        getHeaderViewModel().homeHeaderWalletActionData = null
        view?.updateHeaderItem(getHeaderViewModel())
        getTokocashBalance()
    }

    override fun getShopInfo(url: String, shopDomain: String) {
        getShopInfoByDomainUseCase.execute(GetShopInfoByDomainUseCase.createRequestParam(shopDomain), object : Subscriber<ShopInfo>() {
            override fun onCompleted() {}
            override fun onError(e: Throwable) {
                if (isViewAttached) {
                    view?.openWebViewURL(url)
                }
            }

            override fun onNext(shopInfo: ShopInfo) {
                if (shopInfo.info != null) {
                    view?.startShopInfo(shopInfo.info.shopId)
                } else {
                    view?.openWebViewURL(url)
                }
            }
        })
    }

    override fun openProductPageIfValid(url: String, shopDomain: String) {
        getShopInfoByDomainUseCase.execute(GetShopInfoByDomainUseCase.createRequestParam(shopDomain), object : Subscriber<ShopInfo>() {
            override fun onCompleted() {}
            override fun onError(e: Throwable) {
                if (isViewAttached) {
                    view?.openWebViewURL(url)
                }
            }

            override fun onNext(shopInfo: ShopInfo) {
                if (shopInfo.info != null) {
                    view?.startDeeplinkShopInfo(url)
                } else {
                    view?.openWebViewURL(url)
                }
            }
        })
    }

    override fun getHeaderData(initialStart: Boolean) {
        if (!userSession.isLoggedIn) return
        if (initialStart && headerViewModel != null) {
            if (headerViewModel.homeHeaderWalletActionData == null) getTokocashBalance()
            if (headerViewModel.tokopointsDrawerHomeData == null) getTokopoint()
        } else {
            getTokocashBalance()
            getTokopoint()
        }
    }

    fun hasNextPageFeed(): Boolean {
        return CURSOR_NO_NEXT_PAGE_FEED != currentCursor
    }

    private fun getHeaderViewModel(): HeaderViewModel {
        if (!this::headerViewModel.isInitialized) {
            headerViewModel = HeaderViewModel()
        }
        headerViewModel.isUserLogin = userSession.isLoggedIn
        return headerViewModel
    }

    override fun onDestroy() {
        unsubscribeAllUseCase()
    }

    private fun unsubscribeAllUseCase() {
        if (isActive && !masterJob.isCancelled){
            masterJob.children.map { it.cancel() }
        }
        if (getFeedTabUseCase != null) {
            getFeedTabUseCase?.unsubscribe()
        }
        if (subscription != null) {
            subscription?.unsubscribe()
        }
        if (stickyLoginUseCase != null) {
            stickyLoginUseCase?.cancelJobs()
        }
    }

    /**
     * Tokocash & Tokopoint
     */
    private fun getTokocashBalance() {
            compositeSubscription.add(getWalletBalanceUseCase?.createObservable(RequestParams.EMPTY)
                    .subscribeOn(Schedulers.newThread())
                    .unsubscribeOn(Schedulers.newThread())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(TokocashHomeSubscriber(this)))
        }

    fun getTokocashPendingBalance(){
        compositeSubscription.add(getPendingCasbackUseCase?.createObservable(RequestParams.EMPTY)
                .subscribeOn(Schedulers.newThread())
                .unsubscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(PendingCashbackHomeSubscriber(this)))
    }

    fun getTokopoint(){
        val graphqlResponseObservable = tokopointsObservable
        if (graphqlResponseObservable != null) {
            compositeSubscription.add(graphqlResponseObservable.subscribeOn(Schedulers.newThread())
                    .unsubscribeOn(Schedulers.newThread())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(TokopointHomeSubscriber(this)))
        }
    }

    fun searchHint(){
            val graphqlResponseObservable = keywordSearchObservable
            if (graphqlResponseObservable != null) {
                compositeSubscription.add(graphqlResponseObservable.subscribeOn(Schedulers.newThread())
                        .unsubscribeOn(Schedulers.newThread())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(KeywordSearchHomeSubscriber(this)))
            }
        }

    private val tokopointsObservable: Observable<GraphqlResponse>?
        private get() {
            if (getHomeTokopointsDataUseCaseLazy != null) {
                val getHomeTokopointsDataUseCase = getHomeTokopointsDataUseCaseLazy.get()
                getHomeTokopointsDataUseCase.clearRequest()
                getHomeTokopointsDataUseCase.addRequest(getHomeTokopointsDataUseCase.request)
                return getHomeTokopointsDataUseCase.getExecuteObservable(RequestParams.EMPTY)
            }
            return null
        }

    private val keywordSearchObservable: Observable<GraphqlResponse>?
        private get() {
            if (getKeywordSearchUseCaseLazy != null) {
                val getKeywordSearchUseCase = getKeywordSearchUseCaseLazy?.get()
                getKeywordSearchUseCase.clearRequest()
                getKeywordSearchUseCase.addRequest(getKeywordSearchUseCase.getRequest())
                return getKeywordSearchUseCase.getExecuteObservable(RequestParams.EMPTY)
            }
            return null
        }

    override fun hitBannerImpression(slidesModel: BannerSlidesModel) {
        if (!slidesModel.isImpressed
                && slidesModel.topadsViewUrl != null && !slidesModel.topadsViewUrl.isEmpty()) {
            compositeSubscription.add(Observable.just(ImpresionTask(object : ImpressionListener {
                override fun onSuccess() {
                    slidesModel.isImpressed = true
                }

                override fun onFailed() {
                    slidesModel.isImpressed = false
                }
            }).execute(slidesModel.topadsViewUrl))
                    .debounce(200, TimeUnit.MILLISECONDS)
                    .subscribeOn(Schedulers.newThread())
                    .unsubscribeOn(Schedulers.newThread())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe())
        }
    }

    override fun onBannerClicked(slidesModel: BannerSlidesModel) {
        if (slidesModel.redirectUrl != null && !slidesModel.redirectUrl.isEmpty()) {
            ImpresionTask().execute(slidesModel.redirectUrl)
        }
    }

    override fun updateHeaderTokoPointData(tokopointsDrawerHomeData: TokopointsDrawerHomeData) {
        getHeaderViewModel().setTokoPointDataSuccess()
        getHeaderViewModel().tokopointsDrawerHomeData = tokopointsDrawerHomeData?.tokopointsDrawer
        view?.updateHeaderItem(headerViewModel)
    }

    override fun updateKeywordSearch(keywordSearchData: KeywordSearchData) {
        view?.setHint(keywordSearchData.searchData)
    }

    override fun getFeedTabData() {
        getFeedTabUseCase?.execute(GetFeedTabsSubscriber(view))
    }

    override fun getStickyContent() {
        stickyLoginUseCase?.setParams(StickyLoginConstant.Page.HOME)
        stickyLoginUseCase?.execute(
                { (response) ->
                    for (ticker in response.tickers) {
                        if (ticker.layout == StickyLoginConstant.LAYOUT_FLOATING) {
                            view?.setStickyContent(ticker)
                            return@execute
                        }
                    }
                    view?.hideStickyLogin()
                }
        ) { throwable: Throwable? ->
            view?.hideStickyLogin()
        }
    }

    override fun getPlayBanner(adapterPosition: Int){
        playCardHomeUseCase.execute(
                onSuccess = {
                    view?.setPlayContentBanner(it, adapterPosition)
                },
                onError = {}
        )

    }

    companion object {
        private const val CURSOR_NO_NEXT_PAGE_FEED = "CURSOR_NO_NEXT_PAGE_FEED"
        private var lastRequestTimeHomeData: Long = 0
        private var lastRequestTimeSendGeolocation: Long = 0
        const val FLAG_FROM_NETWORK = 99
        const val FLAG_FROM_CACHE = 98
    }

    init {
        compositeSubscription = CompositeSubscription()
        subscription = Subscriptions.empty()
    }
}
